### PR TITLE
Add Cilium load balancer

### PR DESCRIPTION
**Pull request by iljarotar**
This PR adds a Cilium load balancer as an alternative to metalLB.

### CODE REVIEW

File: metal/cloud.go [hunk 1]
```diff
@@ -92,7 +95,18 @@ func NewCloud(_ io.Reader) (cloudprovider.Interface, error) {
...
+
+	var config loadbalancer.Config
+	switch loadbalancerType {
+	case "metallb":
+		config = metallb.NewMetalLBConfig()
+	case "cilium":
+		config = cilium.NewCiliumConfig()
+	default:
+		return nil, fmt.Errorf("loadbalancer type must be on of 'metallb' or 'cilium'")
+	}
```
Comment (robertvolkmann) [inline metal/cloud.go:107]: For backwards compatibility, the default case should be metallb too.

File: pkg/controllers/loadbalancer/addresspool.go [hunk 2]
```diff
@@ -34,7 +34,7 @@ func (pool *AddressPool) containsCIDR(cidr string) bool {
 	return false
 }
 
-func (pool *AddressPool) appendIP(ip string) {
+func (pool *AddressPool) AppendIP(ip string) {
 	cidr := ip + "/32"
```
Comment (majst01) [inline pkg/controllers/loadbalancer/addresspool.go:38]: This is IPv4 only

File: pkg/controllers/loadbalancer/cilium/config.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+		peer, err := newPeer(n, asn)
+		if err != nil {
+			klog.Warningf("skipping peer: %v", err)
+			continue
+		}
+
+		cfg.Peers = append(cfg.Peers, peer)
+	}
+	return nil
+}
```
Comment (majst01) [inline pkg/controllers/loadbalancer/cilium/config.go:95]: This is duplicate to metallb/config.go

File: pkg/controllers/loadbalancer/cilium/config.go [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+	poolName := fmt.Sprintf("%s-%s", strings.ToLower(network), poolType)
+	pool := cfg.getOrCreateAddressPool(poolName)
+	err := pool.AppendIP(*ip.Ipaddress)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cfg *ciliumConfig) ToYAML() (string, error) {
```
Comment (majst01) [inline pkg/controllers/loadbalancer/cilium/config.go:125]: Only used here, please unexport

File: pkg/controllers/loadbalancer/cilium/config.go [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+				return err
+			}
+		}
+	}
+
+	for _, peer := range cfg.Peers {
+		bgpPeeringPolicy := &ciliumv2alpha1.CiliumBGPPeeringPolicy{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "cilium.io/v2alpha1",
+				Kind:       "CiliumBGPPeeringPolicy",
```
Comment (majst01) [inline pkg/controllers/loadbalancer/cilium/config.go:179]:
```suggestion
				Kind:       ciliumv2alpha1.BGPPKindDefinition,
```

File: pkg/controllers/loadbalancer/cilium/config.go [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+				return err
+			}
+		}
+	}
+
+	for _, pool := range cfg.AddressPools {
+		ipPool := &ciliumv2alpha1.CiliumLoadBalancerIPPool{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "cilium.io/v2alpha1",
+				Kind:       "CiliumLoadBalancerIpPool",
```
Comment (majst01) [inline pkg/controllers/loadbalancer/cilium/config.go:252]:
```suggestion
				Kind:       ciliumv2alpha1.PoolKindDefinition,
```

**Review by majst01 — CHANGES_REQUESTED** [review]

File: pkg/controllers/loadbalancer/cilium/config.go [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, peer := range cfg.Peers {
+		bgpPeeringPolicy := &ciliumv2alpha1.CiliumBGPPeeringPolicy{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "cilium.io/v2alpha1",
```
Comment (robertvolkmann) [inline pkg/controllers/loadbalancer/cilium/config.go:178]:
```suggestion
				APIVersion: ciliumv2alpha1.CustomResourceDefinitionGroup + "/" + CustomResourceDefinitionVersion,
```

File: pkg/controllers/housekeeping/housekeeper.go [hunk 8]
```diff
@@ -85,20 +85,20 @@ func (h *Housekeeper) watchNodes() {
 					return
 				}
 				if oldTunnelAddress == newTunnelAddress {
-					// node was not modified and ip address has not changed, not updating metallb config
+					// node was not modified and ip address has not changed, not updating load balancer config
```
Comment (robertvolkmann) [inline pkg/controllers/housekeeping/housekeeper.go:88]: Inconsistent naming. In some places it is "loadbalancer" and in other places "load balancer".

**Review by robertvolkmann — CHANGES_REQUESTED** [review]

**Review by robertvolkmann — APPROVED** [review]

**Review by Gerrit91 — APPROVED** [review]

**Review by Gerrit91 — APPROVED** [review]

**Review by majst01 — APPROVED** [review]

**Pull request merged**